code/modules/modular_computers/file_system/programs/generic/library.dm

```diff
@@ -0,0 +1,190 @@
+/*
+In reply to this set of comments on lib_machines.dm:
+// TODO: Make this an actual /obj/machinery/computer that can be crafted from circuit boards and such
+// It is August 22nd, 2012... This TODO has already been here for months.. I wonder how long it'll last before someone does something about it.
+
+The answer was five and a half years -ZeroBits
+*/
+
+/datum/computer_file/program/library
+	filename = "library"
+	filedesc = "Library"
+	extended_desc = "This program can be used to view e-books from an external archive."
+	program_icon_state = "word"
+	program_key_state = "atmos_key"
+	program_menu_icon = "note"
+	size = 6
+	requires_exonet = 1
+
+	nanomodule_path = /datum/nano_module/library
+
+/datum/nano_module/library
+	name = "Library"
+	var/error_message = ""
+	var/current_book
+	var/obj/machinery/libraryscanner/scanner
+	var/sort_by = "id"
+
+/datum/nano_module/library/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1, var/datum/topic_state/state = GLOB.default_state)
+	var/list/data = host.initial_data()
+
+	if(error_message)
+		data["error"] = error_message
+	else if(current_book)
+		data["current_book"] = current_book
+	else
+		var/list/all_entries[0]
+		establish_old_db_connection()
+		if(!dbcon_old.IsConnected())
+			error_message = "Unable to contact External Archive. Please contact your system administrator for assistance."
+		else
+			var/DBQuery/query = dbcon_old.NewQuery("SELECT id, author, title, category FROM library ORDER BY "+sanitizeSQL(sort_by))
+			query.Execute()
+
+			while(query.NextRow())
+				all_entries.Add(list(list(
+				"id" = query.item[1],
+				"author" = query.item[2],
+				"title" = query.item[3],
+				"category" = query.item[4]
+			)))
+		data["book_list"] = all_entries
+		data["scanner"] = istype(scanner)
+
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "library.tmpl", "Library Program", 575, 700, state = state)
+		ui.auto_update_layout = 1
+		ui.set_initial_data(data)
+		ui.open()
+
+/datum/nano_module/library/Topic(href, href_list)
```

Comment:
This whole module is going to have problems as the library stuff got removed on Neb in favour of a persistence category.